### PR TITLE
Feature addtg

### PR DESCRIPTION
**Pull request by vanessaxuuan**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import seedu.address.commons.util.CollectionUtil;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.student.Student;
+import seedu.address.model.tutorialgroup.TutorialGroup;
+
+/**
+ * Adds tutorial group to a student identified using it's displayed index from the address book
+ */
+public class AddTutorialGroupCommand extends Command {
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:24]: Perhaps testcases can be added to test the functionality of the AddTutorialGroupCommand?

File: src/main/java/seedu/address/logic/parser/AddTutorialGroupParser.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.commands.AddTutorialGroupCommand;
+import seedu.address.logic.commands.AddTutorialGroupCommand.AddTutorialGroupDescriptor;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.tutorialgroup.TutorialGroup;
+
+/**
+ * Parses input arguments and creates a new AddTutorialGroupCommand object
+ */
+public class AddTutorialGroupParser implements Parser<AddTutorialGroupCommand> {
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/parser/AddTutorialGroupParser.java:18]: Perhaps testcases can be added to test the functionality of the AddTutorialGroupParser?

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    public static final String COMMAND_WORD = "addtg";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds tutorial groups to the student identified "
+            + "by the index number used in the displayed student list. "
+            + "Adding of tutorial groups is cumulative.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_TUTORIAL_GROUP + "TUTORIAL_GROUP]...\n"
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:32]: addTutorialGroup cannot be be optional. Must have at least 1 tg/. 

Should the brackets be there?

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        List<Student> lastShownList = model.getFilteredStudentList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_STUDENT_DISPLAYED_INDEX);
+        }
+        Student studentToEdit = lastShownList.get(index.getZeroBased());
+
+        // identify duplicate modules
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:64]: Should it be "tutorial groups" ?

File: src/main/java/seedu/address/model/student/Student.java [hunk 5]
```diff
@@ -74,6 +74,32 @@ public boolean isSameStudent(Student otherStudent) {
                 && otherStudent.getName().equals(getName());
     }
 
+    /**
+     * Adds additional tutorial group(s) to existing student
+     *
+     * @param toAdd tutorial group(s) to add
+     * @return Student to be updated
+     */
+    public Student addTutorialGroup(Set<TutorialGroup> toAdd) {
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/model/student/Student.java:83]: Student field values are immutable

Should tutorialGroup be modified?

How about consider instantiating a student with the modified tutorialGroups instead?

File: src/main/java/seedu/address/model/student/Student.java [hunk 6]
```diff
@@ -74,6 +74,32 @@ public boolean isSameStudent(Student otherStudent) {
...
+        return this;
+    }
+
+    /**
+     * Checks if the tutorial group already exists
+     *
+     * @param toCheck is a set of tutorial Group(s)
+     * @return if any tutorial group exists under this Student
+     */
+    public boolean tutorialGroupExists(Set<TutorialGroup> toCheck) {
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/model/student/Student.java:94]: Perhaps consider adding testcases to test the functionality of the method?

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 7]
```diff
@@ -68,6 +69,9 @@ public Command parseCommand(String userInput) throws ParseException {
         case HelpCommand.COMMAND_WORD:
             return new HelpCommand();
 
+        case AddTutorialGroupCommand.COMMAND_WORD:
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:72]: Perhaps adding the testcase in AddressBookParser for this new command?

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    /**
+     * Update and returns {@code studentToEdit}
+     * edited with {@code editStudentDescriptor}.
+     */
+    private static Student updateStudent(Student studentToEdit, AddTutorialGroupDescriptor addTutorialGroupDescriptor) {
+        assert studentToEdit != null;
+
+        Set<TutorialGroup> newTutorialGroups = addTutorialGroupDescriptor.getTutorialGroups()
+                .orElse(studentToEdit.getTutorialGroups());
+        return studentToEdit.addTutorialGroup(newTutorialGroups);
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:84]: Student is immutable.

Perhaps create a new Student with the updated tutorial group descriptors?

**Review by LowJiaHao99 — CHANGES_REQUESTED** [review]
Perhaps testcases can be added to check whether the class functions as expected?

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java (hunk 3 above)
Comment (hqhqhq1) [reply src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:32]: For addStudentCommand, I didn't put in brackets for all prefixes, do we need to standardize it?

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        model.setStudent(studentToEdit, updatedStudent);
+        model.updateFilteredStudentList(PREDICATE_SHOW_ALL_STUDENTS);
+        return new CommandResult(String.format(MESSAGE_ADD_TUTORIAL_GROUP_SUCCESS, studentToEdit));
+    }
+
+    /**
+     * Update and returns {@code studentToEdit}
+     * edited with {@code editStudentDescriptor}.
+     */
+    private static Student updateStudent(Student studentToEdit, AddTutorialGroupDescriptor addTutorialGroupDescriptor) {
```
Comment (hqhqhq1) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:79]: maybe can name it as updateStudentTutorialGroup

**Review by hqhqhq1 — COMMENTED** [review]
test cases mentioned by Jia Hao

File: src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        /**
+         * Sets {@code tutorialGroups} to this object's {@code tutorialGroups}.
+         * A defensive copy of {@code tutorialGroups} is used internally.
+         */
+        public void setTutorialGroups(Set<TutorialGroup> tutorialGroups) {
+            this.tutorialGroups = (tutorialGroups != null) ? new HashSet<>(tutorialGroups) : null;
+        }
+
+        /**
+         * Returns an unmodifiable tag set, which throws {@code UnsupportedOperationException}
```
Comment (LowJiaHao99) [inline src/main/java/seedu/address/logic/commands/AddTutorialGroupCommand.java:151]: change tag to tutorial group

File: src/test/java/seedu/address/testutil/AddTutorialGroupDescriptorBuilder.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.stream.Stream;
+
+import seedu.address.logic.commands.AddTutorialGroupCommand;
+import seedu.address.logic.commands.AddTutorialGroupCommand.AddTutorialGroupDescriptor;
+import seedu.address.model.student.Student;
+import seedu.address.model.tutorialgroup.TutorialGroup;
+
+public class AddTutorialGroupDescriptorBuilder {
+
+    public static final AddTutorialGroupDescriptor VALID_DESCRIPTOR_AMY =
```
Comment (LowJiaHao99) [inline src/test/java/seedu/address/testutil/AddTutorialGroupDescriptorBuilder.java:17]: Is VALID_TUTORIAL_GROUP_DESCRIPTOR_AMY more clear?

File: src/test/java/seedu/address/testutil/AddTutorialGroupDescriptorBuilder.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+            new AddTutorialGroupDescriptorBuilder(AMY).build();
+    public static final AddTutorialGroupDescriptor VALID_DESCRIPTOR_BOB =
```
Comment (LowJiaHao99) [inline src/test/java/seedu/address/testutil/AddTutorialGroupDescriptorBuilder.java:19]: Is VALID_TUTORIAL_GROUP_DESCRIPTOR_BOB more clear?

**Review by LowJiaHao99 — CHANGES_REQUESTED** [review]
Minor issues, once resolved LGTM.

**Review by LowJiaHao99 — APPROVED** [review]

**Review by hqhqhq1 — APPROVED** [review]

**Pull request merged**